src/wasm/wasm-binary.cpp

```diff
@@ -1513,11 +1532,122 @@ void WasmBinaryWriter::trackExpressionEnd(Expression* curr, Function* func) {
+
+    // We compute the location of the function declaration area (where the
+    // locals are declared) the first time we need it.
+    BinaryLocation funcDeclarations = 0;
+
+    for (auto& [expr, annotation] : func->codeAnnotations) {
+      if (annotation.branchLikely) {
+        auto exprIter = binaryLocations.expressions.find(expr);
+        if (exprIter == binaryLocations.expressions.end()) {
+          // No expression exists for this annotation - perhaps optimizations
+          // removed it.
+          continue;
+        }
+        auto exprOffset = exprIter->second.start;
+
+        if (!funcDeclarations) {
+          auto funcIter = binaryLocations.functions.find(func.get());
+          assert(funcIter != binaryLocations.functions.end());
+          funcDeclarations = funcIter->second.declarations;
+        }
+
+        // Compute the offset: it should be relative to the start of the
+        // function locals (i.e. the function declarations).
+        auto offset = exprOffset - funcDeclarations;
+
+        funcHints.exprHints.push_back(ExprHint{expr, offset, &annotation});
+      }
+    }
+
+    if (!funcHints.exprHints.empty()) {
+      // We found something. Finalize the data.
+      funcHints.func = func->name;
+
+      // Hints must be sorted by increasing binary offset.
+      std::sort(funcHints.exprHints.begin(),
+                funcHints.exprHints.end(),
+                [](const ExprHint& a, const ExprHint& b) {
+                  return a.offset < b.offset;
+                });
+
+      funcHintsVec.emplace_back(std::move(funcHints));
+    }
+  }
+
+  if (funcHintsVec.empty()) {
+    return {};
+  }
+
+  if (sourceMap) {
+    // TODO: This mode may not matter (when debugging, code annotations are an
+    //       optimization that can be skipped), but atm source maps cause
+    //       annotations to break.
+    Fatal() << "Annotations are not supported with source maps";
+  }
+
+  BufferWithRandomAccess buffer;
+
+  // We found data: emit the section.
+  buffer << uint8_t(BinaryConsts::Custom);
+  auto lebPos = buffer.writeU32LEBPlaceholder();
+  buffer.writeInlineString(Annotations::BranchHint.str);
+
+  buffer << U32LEB(funcHintsVec.size());
+  for (auto& funcHints : funcHintsVec) {
+    buffer << U32LEB(getFunctionIndex(funcHints.func));
+
+    buffer << U32LEB(funcHints.exprHints.size());
+    for (auto& exprHint : funcHints.exprHints) {
+      buffer << U32LEB(exprHint.offset);
+
+      // Hint size, always 1 for now.
+      buffer << U32LEB(1);
+
+      // We must only emit hints that are present.
+      assert(exprHint.hint->branchLikely);
+
+      // Hint contents: likely or not.
+      buffer << U32LEB(int(*exprHint.hint->branchLikely));
+    }
+  }
+
+  // Write the final size. We can ignore the return value, which is the number
+  // of bytes we shrank (if the LEB was smaller than the maximum size), as no
+  // value in this section cares.
+  (void)buffer.emitRetroactiveSectionSizeLEB(lebPos);
```

Comment:
Do we need this void cast to avoid compiler errors? It doesn't seem like we should.